session.go

```diff
@@ -4240,9 +4251,16 @@ func (q *Query) Apply(change Change, result interface{}) (info *ChangeInfo, err
 	session.SetMode(Strong, false)
 
 	var doc valueResult
-	err = session.DB(dbname).Run(&cmd, &doc)
-	if err != nil {
-		if qerr, ok := err.(*QueryError); ok && qerr.Message == "No matching object found" {
+	for i := 0; i < maxUpsertRetries; i++ {
+		err = session.DB(dbname).Run(&cmd, &doc)
+
+		if err == nil {
+			break
+		} else if change.Upsert && IsDup(err) {
```

Comment:
given that the else is redundant, perhaps leave it out and put the if on a new line?